### PR TITLE
Add init keynodes to result structure

### PR DESCRIPTION
**Pull request by SHtress**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation
Adding init keynodes to result struct and tests for this

### CODE REVIEW

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 1]
```diff
@@ -48,22 +48,82 @@ ScAddr ScKeynodes::kBinaryUInt64;
...
+
+  ScAddr resultStruct;
+  bool resultStructValid = SC_FALSE;
+
+  if (result_structure_system_idtf != (sc_char const *)null_ptr)
+  {
+    resultStruct = ctx.HelperResolveSystemIdtf(result_structure_system_idtf, ScType::NodeConstStruct);
+    resultStructValid = resultStruct.IsValid();
+
+    for (ScAddrVector keynodes : keynodeVectors)
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:108]:
```suggestion
    for (ScAddrVector const & keynodes : keynodeVectors)
```

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 2]
```diff
@@ -48,22 +48,82 @@ ScAddr ScKeynodes::kBinaryUInt64;
...
+    {
+      for (ScAddr keynodeAddr : keynodes)
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:110]:
```suggestion
      for (ScAddr const & keynodeAddr : keynodes)
```

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 3]
```diff
@@ -48,22 +48,82 @@ ScAddr ScKeynodes::kBinaryUInt64;
...
+      kBinaryInt16,
+      kBinaryInt32,
+      kBinaryInt64,
+      kBinaryUInt8,
+      kBinaryUInt16,
+      kBinaryUInt32,
+      kBinaryUInt64,
+      kBinaryCustom};
+
+  std::vector<ScAddrVector> const & keynodeVectors = {keynodesAddrs, resultCodes, binaryTypes};
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:98]:
```suggestion
  std::vector<ScAddrVector &> const & keynodeVectors = {keynodesAddrs, resultCodes, binaryTypes};
```

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 4]
```diff
@@ -48,22 +48,82 @@ ScAddr ScKeynodes::kBinaryUInt64;
...
+      }
+    }
+  }
+
   // init sc_result set
-  for (size_t i = 0; i < SC_RESULT_COUNT; ++i)
+  for (ScAddr const & resAddr : resultCodes)
   {
-    ScAddr const resAddr = GetResultCodeAddr(static_cast<sc_result>(i));
     result = result && resAddr.IsValid();
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:120]:
```suggestion
    result &= resAddr.IsValid();
```
else you create tempory object

Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_keynodes.cpp:120]: Check it everywhere and correct

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 5]
```diff
@@ -48,22 +48,82 @@ ScAddr ScKeynodes::kBinaryUInt64;
...
     if (!ctx.HelperCheckEdge(kScResult, resAddr, ScType::EdgeAccessConstPosPerm))
     {
-      result = result && ctx.CreateEdge(ScType::EdgeAccessConstPosPerm, kScResult, resAddr).IsValid();
+      ScAddr const & resEdge = ctx.CreateEdge(ScType::EdgeAccessConstPosPerm, kScResult, resAddr);
+      result = result && resEdge.IsValid();
+      if (result && resultStructValid)
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:125]: So

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 6]
```diff
@@ -75,21 +135,30 @@ bool ScKeynodes::Init(bool force)
...
 
   // command states
-  ScAddr states[] = {kCommandFinishedAddr, kCommandInitiatedAddr, kCommandProgressedAddr};
-  for (auto const & a : states)
+  ScAddrVector const & states = {kCommandFinishedAddr, kCommandInitiatedAddr, kCommandProgressedAddr};
+  for (ScAddr const & a : states)
   {
-    if (!ctx.CreateEdge(ScType::EdgeAccessConstPosPerm, kCommandStateAddr, a).IsValid())
+    ScAddr const & commandStateEdge = ctx.CreateEdge(ScType::EdgeAccessConstPosPerm, kCommandStateAddr, a);
+    if (resultStructValid)
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:147]: Use braced if instead of such it

File: sc-memory/tests/sc-memory/common/sc-memory.cpp [hunk 7]
```diff
@@ -79,3 +79,74 @@ TEST_F(ScMemoryTest, ResolveNodeWithRussianIdtf)
...
+      ScKeynodes::kBinaryInt64,
+      ScKeynodes::kBinaryUInt8,
+      ScKeynodes::kBinaryUInt16,
+      ScKeynodes::kBinaryUInt32,
+      ScKeynodes::kBinaryUInt64,
+      ScKeynodes::kBinaryCustom};
+
+  std::vector<ScAddrVector> const & keynodesVectors = {keynodesAddrs, resultCodes, binaryTypes};
+
+  for (ScAddrVector const & keynodes : keynodesVectors)
```
Comment (NikitaZotov) [inline sc-memory/tests/sc-memory/common/sc-memory.cpp:129]: Divide this cycle into lambdas or functions to improve code readability

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: sc-memory/tests/sc-memory/_test/sc_test.hpp [hunk 8]
```diff
@@ -47,3 +47,29 @@ class ScMemoryTest : public testing::Test
...
+    params.log_level = "Debug";
+
+    params.result_structure_upload = SC_TRUE;
+    params.result_structure_system_idtf = "resultStructure";
+
+    ScMemory::LogMute();
+    ScMemory::Initialize(params);
+    ScMemory::LogUnmute();
+  }
+};
```
Comment (NikitaZotov) [inline sc-memory/tests/sc-memory/_test/sc_test.hpp:75]:
```suggestion
};

```

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 9]
```diff
@@ -75,21 +135,37 @@ bool ScKeynodes::Init(bool force)
     if (!item.IsValid())
       result = false;
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:136]: Check it logic please

File: sc-memory/tests/sc-memory/common/sc-memory.cpp [hunk 10]
```diff
@@ -2,9 +2,28 @@
...
 
 #include "sc_test.hpp"
 
+void checkConnectionInStruct(
+    ScMemoryContext * m_ctx,
+    ScAddr const & keynodeAddr,
+    ScAddr const & otherKeynodeAddr,
+    ScAddr const & resultStruct)
+{
+  ScIterator3Ptr it3 = m_ctx->Iterator3(keynodeAddr, ScType::EdgeAccessConstPosPerm, otherKeynodeAddr);
```
Comment (NikitaZotov) [inline sc-memory/tests/sc-memory/common/sc-memory.cpp:15]:
```suggestion
  ScIterator3Ptr const it3 = m_ctx->Iterator3(keynodeAddr, ScType::EdgeAccessConstPosPerm, otherKeynodeAddr);
```

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

File: config/sc-machine.ini [hunk 11]
```diff
@@ -12,6 +12,9 @@ log_type = Console
 log_file = ../sc-memory.log
 log_level = Info
 
+result_structure_system_idtf = resultStructure
+result_structure_upload = true
```
Comment (smousy) [inline config/sc-machine.ini:16]: maybe duplicated is not a good idea? What if memory result_structure_system_idtf != builder  result_structure_system_idtf ? should stay parameters in builder section or different names for parameters in memory section

**Review by smousy — CHANGES_REQUESTED** [review]

File: sc-memory/tests/sc-memory/common/sc-memory.cpp [hunk 12]
```diff
@@ -79,3 +98,66 @@ TEST_F(ScMemoryTest, ResolveNodeWithRussianIdtf)
   EXPECT_TRUE(m_ctx->HelperSetSystemIdtf(englishIdtf, englishNode));
   EXPECT_EQ(m_ctx->HelperFindBySystemIdtf(englishIdtf), englishNode);
 }
+
+TEST_F(ScMemoryTestWithResultStruct, TestResultStruct)
```
Comment (KovalM) [inline sc-memory/tests/sc-memory/common/sc-memory.cpp:102]: This test should also check that link with system idtf of keynode, edges to this links, also was added to result.

test_keynode
 => [test_keynode]
  ^
  |
 nrel_system_identificator

File: sc-memory/sc-memory/sc_memory.cpp [hunk 13]
```diff
@@ -89,7 +89,7 @@ bool ScMemory::Initialize(sc_memory_params const & params)
   if (ms_globalContext == nullptr)
     return false;
 
-  ScKeynodes::Init();
+  ScKeynodes::Init(false, params.result_structure_upload ? params.result_structure_system_idtf : (sc_char *)null_ptr);
```
Comment (KovalM) [inline sc-memory/sc-memory/sc_memory.cpp:92]: For discuss: what about keynodes from applications

**Review by KovalM — CHANGES_REQUESTED** [review]

**Review by KovalM — APPROVED** [review]

File: sc-memory/sc-core/sc_memory_params.c [hunk 14]
```diff
@@ -30,4 +30,7 @@ void sc_memory_params_clear(sc_memory_params * params)
 
   params->max_loaded_segments = DEFAULT_MAX_LOADED_SEGMENTS;
   params->max_threads = DEFAULT_MAX_THREADS;
+
+  params->result_structure_system_idtf = (sc_char const *)null_ptr;
```
Comment (NikitaZotov) [inline sc-memory/sc-core/sc_memory_params.c:34]: Change result_structure by init_memory_generated_structure

File: sc-memory/sc-memory/sc_keynodes.cpp [hunk 15]
```diff
@@ -74,22 +134,38 @@ bool ScKeynodes::Init(bool force)
...
   // binary types
+  for (ScAddr const & binaryType : binaryTypes)
   {
-    ScSet set(ctx, kBinaryType);
-    set << kBinaryDouble << kBinaryFloat << kBinaryString << kBinaryInt8 << kBinaryInt16 << kBinaryInt32 << kBinaryInt64
-        << kBinaryUInt8 << kBinaryUInt16 << kBinaryUInt32 << kBinaryUInt64 << kBinaryCustom;
+    if (!ctx.HelperCheckEdge(kBinaryType, binaryType, ScType::EdgeAccessConstPosPerm))
+    {
+      ScAddr const & binaryTypeEdge = ctx.CreateEdge(ScType::EdgeAccessConstPosPerm, kBinaryType, binaryType);
+      if (resultStructValid && binaryTypeEdge.IsValid())
```
Comment (NikitaZotov) [inline sc-memory/sc-memory/sc_keynodes.cpp:163]:
```suggestion
      if (initMemoryGeneratedStructureValid && binaryTypeEdge.IsValid())
```

File: sc-tools/sc-builder/src/main.cpp [hunk 16]
```diff
@@ -48,9 +48,14 @@ try
   ScParams memoryParams{options, keys};
   memoryParams.insert({"repo_path", options[{"output_path", "o"}].second});
 
-  ScMemoryConfig memoryConfig{config, {"repo_path", "log_file"}, std::move(memoryParams)};
   ScBuilderConfig builderConfig{ScConfig(config, {"repo_path", "log_file"}), std::move(params)};
 
+  memoryParams.insert(
```
Comment (NikitaZotov) [inline sc-tools/sc-builder/src/main.cpp:53]: Fix namings here

**Review by NikitaZotov — CHANGES_REQUESTED** [review]

**Review by NikitaZotov — APPROVED** [review]

**Pull request merged**